runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java

```diff
@@ -215,22 +221,76 @@ public static Environment createProcessEnvironment(
     }
 
     ImmutableList.Builder<ArtifactInformation> filesToStage = ImmutableList.builder();
+    ImmutableMap.Builder<String, Function<ArtifactInformation, ArtifactInformation>>
+        lazyArtifactsBuilder = ImmutableMap.builder();
     for (String path : pathsToStage) {
       File file = new File(path);
-      if (new File(path).exists()) {
-        // Spurious items get added to the classpath. Filter by just those that exist.
-        if (file.isDirectory()) {
-          // Zip up directories so we can upload them to the artifact service.
-          try {
-            filesToStage.add(createArtifactInformation(zipDirectory(file)));
-          } catch (IOException e) {
-            throw new RuntimeException(e);
-          }
-        } else {
-          filesToStage.add(createArtifactInformation(file));
-        }
+      // Spurious items get added to the classpath. Filter by just those that exist.
+      if (file.exists()) {
+        String id = UUID.randomUUID().toString();
```

Comment:
Note that you could put one ID into the map for the entire list of files if you allowed ResolutionFn to return a List/Collection of artifacts.

Reply:
Yes, I think the goal is to avoid even this enumeration until we actually need it.

Reply:
done.